README.md
Система рассылки уведомлений через IM с API-сервисом.

On english:
System for sending notifications via chat with the API service.

** Im / im im (matrix)**
Notification distribution system.

The logic of the system is as follows:

- Im im (matrix)
- users are connected to it, im im (matrix)
- in the system, they also have a user on whose behalf the mailing is carried out (@bot:corp.ru)
- The author of any system that should be used for mailing takes the API key from the system administrator of the mailing service
- and configures its system to send JSON data (one of the fields must contain the previously received API key) with the information to be sent. Data in the JSON text format must contain the parcel and the recipient's address.
- the recipient's address can consist of two fields: email address and email address.
- the logic for sending mailings to these addresses is as follows: first, it tries to send a message via chat, but if you can't send it within 20 minutes, we send it by mail. If the address is not specified in principle, it is sent to the mail immediately. Attempts to send by mail servsi undertakes within 3 hours. If you can't send it, failed "failed".
- im im (matrix) - marked in the mailing status for this message
- each status change in the mailing list is marked -" in progress", "sent", "sent", "read", etc.
- when applying for the newsletter requesting system may transmit the field "callback_url" containing a link at which the mailing system will notify you of changes by sending this message sending this link to JSON data with the information in this newsletter
- when prompted for a newsletter, requesting a service may pass as your internal identifier of this message (it is stored in the database of mailing and will return when callback)
- call back s can be somewhat. Call callback-a for this message with all statuses on it.

**Sending**

Input:
* api_key** - API access key (string, required parameter-request from the API service hoster)
* **address_im** - to the recipient's address in the service (string, must be specified, or must be specified address_email, or both)
* *address_email** - recipient's email address (string, must be specified, add address_im, or both)
* **message** - message text
* **type** - message text type (string, optional parameter. Possible values: HTML, text)
* **callback_url** - web address of the callback (string, optional parameter. The service will send the message distribution status to this address when it is updated) - J JSON
* **sender_uniq_id** - unique ID of sending (string, optional parameter. This ID is saved in the mailing list database and you can use it to get the message's mailing status)
* **edit_previous** - optional option (false by default) that enables editing of previous sent messages with the same sender_uniq_id. (value: TRUE/false). If true - if the database has a previous successfully sent message with the same sender_uniq_id, then instead of sending a new message, the system will correct the text of the previous message in the chat room to a new one. In this case, the chat will show the PostScript "edited". This message will also have the status * * edited_previous_sended_by_im**

Output:
* **message_id** - internal ID of the message in the queue to be sent (integer)
* **status** - sending status (string, "success" error "error")
* **description** - detailed description in case of an error (string)

Request example:

By this URL:

https://api-sending-messages.corp.ru/api_add_message.cgi

we send it in JSON in the form:
```
{'address_im':'@user:matrix-server.com', 'message':'test message', 'api_key':'secret_api_key'}
```
Or:
```
{'address_email':'user@mail.com', 'message':'test message', 'api_key':'secret_api_key'}
```
Or:
```
{
'address_im': '@user:matrix-server.com',
'address_email':'user@mail.com',
'message':' < P>to test in HTML message: < / P><ol><Li>first< / Li><Li>second< / Li><Li>third< / Li>\n< / PR>',
'Type': 'HTML code',
'api_key': 'secret_api_key',
'sender_uniq_id': '23443',
'callback_url':'http://my-server.com/status_sending.php'
}
```


In response, the server will send the message ID, J JSON:
```
{
'message_id': 23,
"status": "success",
'description': 'add a message to the sending queue'
}
```

If an error occurs:
```
{
"status": "error",
"description": "some mistake",
}
```
** Receiving notification of the sending status **

Url "callback" url-and the service will send (several times) a JSON containing updated information on the status of sending and reading the message.

Data in JSON that the service will send:
* id-internal ID of the message in the queue to be sent (integer)
* sending_status - sending status (string, see below for status options)
* address_im-im im
* address_email-recipient's email Address
* time_create-time when the message was generated
* time_start_process-time when the sender started processing the message
* time_send-time when the message was sent
* time_read-time when the recipient reads the message
* sending_retry_num-Number of attempts to send the message
* sender_uniq_id-ID Passed from the sender to identify the message
* error_description_im-Text description of the error in sending them
* error_description_email-Text description of the email error
* callback_status-callback Status of sending information about the mailing list by
* callback_retry_num-callback Number of attempts to send information by
* im_event_id - Internal ID of the message in the instant messaging system

The status of sending a message ('sending_status') can take the following values:
* "new" - a New message to send
* "no_room" - you Need to create a room to communicate with the user
* "in progress" - in the process of sending
* 'sending_by_im' - im im
* 'sending_by_email' - Sent by mail
* 'sent_by_im' - im im
* 'sent_by_email' - Sent by mail
* "read" - Read by the recipient
* "error" - sending Failed, but attempts to resend continue
* "error" - sending Failed, attempts to resend stopped

The status of sending a callback ('callback_status') can take the following values:
* 'new' - ' Newly created request -back callback
* "need to send" - the status of sending the message has been updated. You must notify the recipient via a callback
* 'sent' - the status of sending the message was successfully sent to the recipient via collback
* "error" - sending Failed, but attempts to resend continue
* "error" - sending Failed, attempts to resend stopped


Example of the received JSON:

Successful submission:
```
{
"address_email": null,
"address_im": "@user:corp.ru",
"callback_retry_num": 3,
"callback_status": "error",
"error_description_email": null,
"error_description_im": null,
"im_event_id": "$7gFwqkL324525433WEw2pUyHzwhZ1XMAyrzgjpd6aek",
"message_id": 3,
"sender_uniq_id": null,
"sending_retry_num": 1,
"sending_status": "read",
"time_create": "2020-05-21 10: 46:40.605750",
"time_read": "2020-05-21 15: 03:14",
"time_send": "2020-05-21 10: 47: 01",
"time_start_process": "2020-05-21 10: 46:59"
}
```
Complete sending failure:
```
{
"address_email": null,
"address_im": "@user:corp.ru",
"callback_retry_num": 0,
"callback_status": "error",
"error_description_email": null,
"error_description_im": "im im im-contact the administrator",
"im_event_id": null,
"message_id": 1,
"sender_uniq_id": null,
"sending_retry_num": 1,
"sending_status": "error",
"time_create": "2020-05-20 19: 08:12.899453",
"time_read": zero,
"time_send": "2020-05-20 19: 08: 50",
"time_start_process": "2020-05-20 19: 08:49"
}
```

НА русском:

**Рассылка уведомлений через почту и/или IM (Matrix)**
Система рассылки уведомлений.

Логика работы системы следующая:

  - Есть сервер IM (matrix)
  - к нему подключены пользователи, у них есть адреса в системе IM (matrix)
  - в системе IM так же заведён пользователь, от имени которого проводится рассылка (@bot:corp.ru)
  - Автор какой-либо системы, с которой необходимо проводить рассылку берёт у системного администратора сервиса рассылки API-ключ
  - и настраивает свою систему, чтобы она по необходимости рассылки - отправляла json-данные (одно из полей должно содержать полученный ранее API-ключ) с информацией, которую нужно отправить. Данные json должны содержать текст посылки, адрес получателя.
  - адрес получателя может состоять из двух полей - IM-адреса и почтового.
  - логика отправки у рассылки по этим адресам следующая: сначала пытается отправить сообщение через IM, если же в течении 20 минут не получилось отправить - отправляем по почте. Если же адрес IM в принципе не указан, то на почту отправляется сразу. Попытки отправить по почте сервси предпринимает в течении 3-х часов. Если же отправить не удалось, то статус у этого запроса приобретает наименование 'failed'.
  - ошибки отправки по почте и через IM (matrix) - помечаются в статусе рассылки для этого сообщения
  - каждое изменение статуса в рассылке сообщения помечается - "в процессе", "отправляется", "отправилось", "прочитано" и т.п.
  - при подаче заявки на рассылку запрашивающая система может передать поле "callback_url", содержащее ссылку, по которой система рассылки будет уведомлять об изменениях по рассылке данного сообщения, отправляя по этой ссылке json-данные с информацией по данной рассылке
  - при запросе на рассылку, запрашивающий сервис может передать так же свой внутренний идентификатор этого сообщения (он сохранится в базе рассылки и вернётся при callback)
  - callback-ов может быть несколько. Каждое изменение статуса рассылки провоцирует отправку callback-а для данного сообщения со всеми статусами по нему.

**Отправка**

Входные данные:
  * **api_key** - ключ доступа к АПИ (строка, обязательный параметр - запросить у хостера АПИ-сервиса)
  * **address_im** - адрес получателя в IM сервисе (строка, должен быть указан, либо должен быть указан address_email, либо оба)
  * **address_email** - почтовый адрес получателя (строка, должен быть указан, либо должен быть указан address_im, либо оба)
  * **message** - текст сообщения
  * **type** - тип текста сообщения (строка, необязательный параметр. Возможные значения: html, text)
  * **callback_url** - web-адрес обратного вызова (строка, необязательный параметр. По данному адресу сервис будет отправлять статус рассылки сообщения при его обновлении) - отправляться будет json
  * **sender_uniq_id** - уникальный идентификатор отправки (строка, необязательный параметр. Этот идентификатор сохраняется в базе рассылки и по нему можно получить статус рассылки сообщения)
  * **edit_previouse** - необязательная опция (по-умолчанию - False), включающая редактирование предыдущих отправленных сообщений с таким же sender_uniq_id. (значение: True/False). В случае True - если в базе есть предыдущее успешно отправленное сообщение с таким же sender_uniq_id, то вместо отправки нового сообщения - система исправит текст предыдущего сообщения в комнате чата на новый. При этом чат покажет приписку "отредактировано". У этого сообщения же станет статус **edited_previous_sended_by_im**

Выходные данные:
  * **message_id** - внутренний идентификатор сообщения в очереди на отправку (целое число)
  * **status** - статус отправки (строка, "success" или "error")
  * **description** - детальное описание в случае ошибки (строка)

Пример запроса:

По данному url:

https://api-sending-messages.corp.ru/api_add_message.cgi

отпраляем json вида:
```
{'address_im':'@user:matrix-server.com','message':'test message', 'api_key':'secret_api_key'}
```
Или:
```
{'address_email':'user@mail.com','message':'test message', 'api_key':'secret_api_key'}
```
Или:
```
{
  'address_im':'@user:matrix-server.com',
  'address_email':'user@mail.com',
  'message':'<p>test html-message: </p><ol><li>первое</li><li>второе</li><li>третье</li>\n</ol>',
  'type':'html',
  'api_key':'secret_api_key',
  'sender_uniq_id':'23443',
  'callback_url':'http://my-server.com/status_sending.php'
}
```

 
В ответ сервер пришлёт идентификатор сообщения, в виде json:
```
{
  'message_id':23,
  'status': 'success',
  'description': 'add message to sending queue'
}
```

В случае ошибки:
```
{
  "status": "error",
  "description": "some error",
}
```
** Получение уведомления о статусе отправки **

По указанному в запросе url "callback"-а сервис отправит (несколько раз) json, содержащий обновляему информацию по статусу отправки и прочтения сообщения.

Данные в json, который отправит сервис:
  * id - внутренний идентификатор сообщения в очереди на отправку (целое число)
  * sending_status - статус отправки (строка, варианты статусов см. ниже)
  * address_im - Адрес пользователя в IM
  * address_email - Адрес электронной почты получателя
  * time_create - Время формирования сообщения
  * time_start_process - Время начала обработки сообщения отправщиком
  * time_send - Время отправки сообщения
  * time_read - Время прочтения сообщения получателем
  * sending_retry_num - Количество попыток отправки сообщения
  * sender_uniq_id - Идентификатор, переданный от отправителя, для идентификации сообщения
  * error_description_im - Текстовое описание ошибки отправки по IM
  * error_description_email - Текстовое описание ошибки отправки по почте
  * callback_status - Статус отправки информации о рассылке по callback
  * callback_retry_num - Количество попыток отправок информации по callback
  * im_event_id - Внутренний идентификатор сообщения в системе IM

Статус отправки сообщения ('sending_status') может принимать значения:
  * 'new' - Новое сообщение для отправки
  * 'no_room' - Необходимо создать комнату для общения с пользователем
  * 'in progress' - В процессе отправки
  * 'sending_by_im' - Отправляется через IM
  * 'sending_by_email' - Отправляется по почте
  * 'sent_by_im' - Отправлено через IM
  * 'sent_by_email' - Отправлено по почте
  * 'readed' - Прочитано получателем
  * 'error' - Сбой отправки, но попытки повторной отправки продолжаются
  * 'fault' - Сбой отправки, попытки повторной отправки прекращены

Статус отправки callback ('callback_status') может принимать значения:
  * 'new' - 'Вновь созданный запрос - пока нет необходимости слать информацию по callback
  * 'need send' - Статус отправки сообщения обновлён. Необходимо уведомить об этом получаетя через callback
  * 'sended' - Статус отправки сообщения успешно отправлен получателю по collback
  * 'error' - Сбой отправки, но попытки повторной отправки продолжаются
  * 'fault' - Сбой отправки, попытки повторной отправки прекращены


Пример получаемого json:

Успешная отправка:
```
 {
    "address_email": null,
    "address_im": "@user:corp.ru",
    "callback_retry_num": 3,
    "callback_status": "error",
    "error_description_email": null,
    "error_description_im": null,
    "im_event_id": "$7gFwqkL324525433WEw2pUyHzwhZ1XMAYRZgJpd6aEk",
    "message_id": 3,
    "sender_uniq_id": null,
    "sending_retry_num": 1,
    "sending_status": "readed",
    "time_create": "2020-05-21 10:46:40.605750",
    "time_read": "2020-05-21 15:03:14",
    "time_send": "2020-05-21 10:47:01",
    "time_start_process": "2020-05-21 10:46:59"
}
```
Полный сбой отправки:
```
{
    "address_email": null,
    "address_im": "@user:corp.ru",
    "callback_retry_num": 0,
    "callback_status": "error",
    "error_description_email": null,
    "error_description_im": "ошибка заведения комнаты в IM для посылки сообщения - обратитесь к администратору",
    "im_event_id": null,
    "message_id": 1,
    "sender_uniq_id": null,
    "sending_retry_num": 1,
    "sending_status": "fault",
    "time_create": "2020-05-20 19:08:12.899453",
    "time_read": null,
    "time_send": "2020-05-20 19:08:50",
    "time_start_process": "2020-05-20 19:08:49"
}
```




Changelog.txt
Версии можно посмотреть через запуск команды: 'git tag'

Версия 1.2
1. Добавил поддержку правки предыдущих сообщений

Версия 1.1 
1. Поправил ошибку падения в случае пустых полей в базе
2. Бот умеет статистику по отправкам и слать несколько отчётов в экселе/html


message_sender/api_callback_requests.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import traceback
import sys
import time
import datetime
import psycopg2
import psycopg2.extras
import os
import logging
import json
from logging import handlers
import config as conf
import urllib
import requests

global_error_descr=""
conn=None
cur=None
log=None

def send_callback_json(callback_url,json_data):
  global log
  log.debug("try send to '%s')"%callback_url)
  time_execute=time.time()
  num=0
  try:
    response = requests.post(callback_url, json=json_data)
    data = response.raw      # a `bytes` object
    if response.status_code != 200:
      log.error("response != 200 (OK)")
      return False
  except:
    log.error("post data to url: %s"%callback_url)
    return False
  log.debug("execute function time=%f"%(time.time()-time_execute))
  return True
   
def set_callback_retry_num(message_id, callback_retry_num):
  global conn
  global cur
  global log
  time_execute=time.time()
  try:
    sql="update tbl_sending_queue set \
      callback_retry_num=%(callback_retry_num)d"\
      % {"callback_retry_num":callback_retry_num}
    sql=sql+" where id=%d" % message_id
  except Exception as e:
    log.error(get_exception_traceback_descr(e))
    return False

  log.debug("sql=%s" % sql)
  try:
    cur.execute(sql)
    conn.commit()
    log.debug("Row(s) were updated : %s"%str(cur.rowcount))
  except psycopg2.Error as e:
    log.error("can not update row: %s" % e.pgerror)
    log.info("try rollback update for this connection")
    try:
      conn.rollback()
    except psycopg2.Error as e:
      log.error("sql error: %s" % e.pgerror)
    return False

  log.debug("execute function time=%f"%(time.time()-time_execute))
  return True

def set_callback_status(message_id, callback_status):
  global conn
  global cur
  global log
  time_execute=time.time()
  try:
    sql="update tbl_sending_queue set \
      callback_status_id=(select callback_status_id from tbl_callback_status where callback_status_name='%(callback_status)s')"\
      % {"callback_status":callback_status}
    sql=sql+" where id=%d" % message_id
  except Exception as e:
    log.error(get_exception_traceback_descr(e))
    return False

  log.debug("sql=%s" % sql)
  try:
    cur.execute(sql)
    conn.commit()
    log.debug("Row(s) were updated : %s"%str(cur.rowcount))
  except psycopg2.Error as e:
    log.error("can not update row: %s" % e.pgerror)
    log.info("try rollback update for this connection")
    try:
      conn.rollback()
    except psycopg2.Error as e:
      log.error("sql error: %s" % e.pgerror)
    return False

  log.debug("execute function time=%f"%(time.time()-time_execute))
  return True

def process_needed_callbacks():
  global log
  global conn
  global cur
  global global_error_descr

  time_execute=time.time()
  try:
    # формируем sql-запрос:
    sql="""
    select 
      a.id,
      a.address_im,
      a.address_email,
      a.time_create,
      a.time_start_process,
      a.time_send,
      a.time_read,
      b.sending_status_name,
      a.sending_retry_num,
      a.sender_uniq_id,
      a.error_description_im,
      a.error_description_email,
      c.callback_status_name,
      a.callback_retry_num,
      a.im_event_id,
      a.callback_url,
      a.edit_previouse
      from tbl_sending_queue as a
      LEFT JOIN tbl_sending_status as b ON a.sending_status_id = b.sending_status_id
      LEFT JOIN tbl_callback_status as c ON a.callback_status_id = c.callback_status_id
      where a.callback_url is not null and a.callback_status_id in (select c.callback_status_id from tbl_callback_status where c.callback_status_name='need send' or callback_status_name='error');
    """

    try:
      cur.execute(sql)
      items = cur.fetchall()
    except psycopg2.Error as e:
      log.error("I am unable select data from tbl_sending_queue: %s" % e.pgerror)
      return False
    log.debug("execute select() time=%f"%(time.time()-time_execute))
    # обработка:
    for item in items:
      time_proccess_one_callback=time.time()
      log.info("proccess callback for message id=%d"%item[0])
      data={}
      data["message_id"]=item[0]
      data["address_im"]=item[1]
      data["address_email"]=item[2]
      if item[3]!=None:
        data["time_create"]=str(item[3])
      else:
        data["time_create"]=None
      if item[4]!=None:
        data["time_start_process"]=str(item[4])
      else:
        data["time_start_process"]=None
      if item[5]!=None:
        data["time_send"]=str(item[5])
      else:
        data["time_send"]=None
      if item[6]!=None:
        data["time_read"]=str(item[6])
      else:
        data["time_read"]=None
      data["sending_status"]=item[7]
      data["sending_retry_num"]=item[8]
      data["sender_uniq_id"]=item[9]
      data["error_description_im"]=item[10]
      data["error_description_email"]=item[11]
      data["callback_status"]=item[12]
      data["callback_retry_num"]=item[13]
      data["im_event_id"]=item[14]
      callback_url=item[15]
      data["edit_previouse"]=item[16]

      if data["callback_status"] == "error" and data["callback_retry_num"] == conf.callback_max_retry:
        log.warning("callback_max_retry (%d) exeed for callback() of message_id=%d - set callback status as 'fault' for item"%(conf.callback_max_retry, data["message_id"]))
        if set_callback_status(data["message_id"],"fault") == False:
          log.error("set_callback_status()")
          return False
        continue

      # отправляем 
      log.debug(json.dumps(data, indent=4, sort_keys=True,ensure_ascii=False))
      if send_callback_json(callback_url,data) == False:
        # не получилось отправить callback. Помечаем статус отправки callback-а:
        log.warning("send_callback_json('%s')"%callback_url)
        if set_callback_status(data["message_id"],"error") == False:
          log.error("set_callback_status()")
          return False
        if set_callback_retry_num(data["message_id"],data["callback_retry_num"]+1) == False:
          log.error("set_callback_retry_num()")
          return False
      else:
        # успешно отправили callback, сбрасываем счётчик ошибок отправок, если нужно:
        if data["callback_retry_num"] > 0:
          if set_callback_retry_num(data["id"],0) == False:
            log.error("set_callback_retry_num()")
            return False
        # помечаем как отправленное:
        if set_callback_status(data["id"],"sended") == False:
          log.error("set_callback_status()")
          return False
      log.debug("time proccess one callback=%f"%(time.time()-time_proccess_one_callback))

  except Exception as e:
    log.error(get_exception_traceback_descr(e))
    return False
  log.debug("execute function time=%f"%(time.time()-time_execute))
  return True
  

def get_exception_traceback_descr(e):
  if hasattr(e, '__traceback__'):
    tb_str = traceback.format_exception(etype=type(e), value=e, tb=e.__traceback__)
    result=""
    for msg in tb_str:
      result+=msg
    return result
  else:
    return e

def connect_to_db():
  global conn
  global cur
  try:
    log.debug("connect to: dbname='" + conf.send_db_name + "' user='" +conf.send_db_user + "' host='" + conf.send_db_host + "' password='" + conf.send_db_passwd + "'")
    conn = psycopg2.connect("dbname='" + conf.send_db_name + "' user='" +conf.send_db_user + "' host='" + conf.send_db_host + "' password='" + conf.send_db_passwd + "'")
    cur = conn.cursor()
  except Exception as e:
    log.error(get_exception_traceback_descr(e))
    log.error("I am unable to connect to the database")
    return False
  return True

#=============== main() ===============
def main():
  global log
  if connect_to_db()==False:
    log.error("connect_to_db()")
    return False
  if process_needed_callbacks() == False:
    log.error("process_needed_callbacks()")
    return False
  return True

if __name__ == '__main__':
  log=logging.getLogger("api_callback_requests")
  if conf.debug:
    log.setLevel(logging.DEBUG)
  else:
    log.setLevel(logging.INFO)

  # create the logging file handler
  fh = logging.handlers.TimedRotatingFileHandler(conf.log_path_api_callback_requests, when=conf.log_backup_when, backupCount=conf.log_backup_count, encoding='utf-8')
  formatter = logging.Formatter('%(asctime)s - %(name)s - %(filename)s:%(lineno)d - %(funcName)s() %(levelname)s - %(message)s')
  fh.setFormatter(formatter)

  if conf.debug:
    # логирование в консоль:
    #stdout = logging.FileHandler("/dev/stdout")
    stdout = logging.StreamHandler(sys.stdout)
    stdout.setFormatter(formatter)
    log.addHandler(stdout)

  # add handler to logger object
  log.addHandler(fh)

  log.info("Program started")

  log.info("python version=%s"%sys.version)

  if main()==False:
    log.error("error main()")
    sys.exit(1)
  log.info("Program exit!")
